### DISCUSSION TITLE
Unbundle kernel+initrd from images

### DISCUSSION THREAD

**Discussion by Itxaka** [rancher/elemental-cli#344] [Ideas]
Current:
 - kernel+initrd are bundled inside each image
 - grub needs to load extra modules to support mounting those images (loop+squash)
 - kernel upgrade means full image upgrade
 - secureboot depends on grub provided the modules as part of the signed grub
 - 3 different places for bootargs (active,passive,recovery)

Poposal:
 - Extract kernel+initrd to state partition
 - grub.cfg still chainloads the grub.cfg in state partition
 - no extra modules loaded at grub time as we dont need to open the images to boot the kernel
 - secureboot out of the box with this implementation
 - only one place for kernel+initrd+bootargs
 - simple install, simpler upgrade
 - separation of kernel and system, future improvement to install/upgrade (smaller images, separated images, possibility of upgrading one or the other?)

grub.efi boots -> loads grub.cfg on EFI partition -> mounts state dir -> loads grub.cfg+bootargs+extra grub files from state -> loads kernel+initrd from state partition -> boots kernel with proper args (LABEL, cos-img/filename) -> Boots system

This has been tested by modifying the grub files and kernel+initrd and worked under secureboot, so its doable.

Cons:
 - extend elemental-cli to move the files into the state partition
 - change grub packages to have different config files
 - would need to extend the install/upgrade to also be able to only upgrade kernel or system
 - users need to create 2 images if they want to take advantage of smaller downloads and different upgrade targets
 - system images would still contain the kernel+initrd, removing it to make them smaller would require change to your image build
 - could also not change the images so they are the same as before but we would then have extra files around (kernel+initd in active, passive, recovery AND state partiton....what a mess!)
 - could also remove the kernel+initrd from the images as part of install/upgrade, but no advantages on image sizes that way.
 - possible BRAKING CHANGE to build, images, operator, teal, elemental-cli, everything :/

**Comment by Itxaka** [comment]
cc @mudler @davidcassany @fgiudici @frelon @alexander-demicev @agracey

**Comment by mudler** [comment]
my cents on this is that this means to go back at:
- Not being a real single-image, atomic upgrade. Now you have also kernel and initrd as moving pieces to take into consideration. Next will be grub?
- Add runtime operations that can break upgrades - the initial design goal is to avoid as many moving pieces as possible to have a clean 1-1 swap during upgrades and minimize the impact of a failure during upgrade. With the kernel operation included too, now you have an unknown state where the machine can be if powered off during the upgrade. Note that the current approach is atomic as nothing changes  and its an atomic action - or you upgrade, or you are back with the old system.

to me here the project should take an explicit call on those topics. my guess is that it would be better to start from scratch with another design if you are tailored to specific requirements where this design doesn't fit. however - you can't have pros from both worlds.

**Reply by Itxaka** [reply]
> my cents on this is that this means to go back at:
> 
>     * Not being a real single-image, atomic upgrade. Now you have also kernel and initrd as moving pieces to take into consideration. Next will be grub?
> 
>     * Add runtime operations that can break upgrades - the initial design goal is to avoid as many moving pieces as possible to have a clean 1-1 swap during upgrades and minimize the impact of a failure during upgrade. With the kernel operation included too, now you have an unknown state where the machine can be if powered off during the upgrade. Note that the current approach is atomic as nothing changes  and its an atomic action - or you upgrade, or you are back with the old system.
> 

This is a very good point but its also very simply solved, nothing prevents us for having the old kernel and new kernel side to side, so you upgrade the system and machine is powered off during the kernel "upgrade", you still boot with the old one. The only sensible point is the grub.cfg update, where it can be in a broken state when updating but this could also be worked around by having the same thing as with A/B image, you get an A/B kernel :D

Still, a very good point and this adds too much complexity IMHO. 3 boot images * 2 kernels means 6 options. Which is doable with grub submenus and implemented in other systems (btrfs comes to mind, you get btrfs snapshots entries to boot from). But the work to support this properly is a big change on itself, which may not be worth it.

**Comment by davidcassany** [comment]
My feeling on that is this is adding OS management complexity to essentially simplify grub2 module loading. I see and share your the point of limiting as much as possible the grub2 attributions and I also agree our grub2 setup is not perfect. However I am not sure this options is breaking too much the spirit of containerized OSes. Because of this systemd-boot was not considered at a time, because it requires Kernel and Initrd inside the EFI partition.

To my eyes having to extract parts of the OS image separately is going to the direction of regular OS management (many small pieces).

In order to avoid loopback devices (some more wild thinking 😅 ) we could embrace volumes (lvm) and subvolumes (btrfs), but this just shifts the grub2 loopback module needs to lvm module or btrfs module... I wouldn't be surprised lvm and btrfs modules are properly maintained and included in signed grub2 efi images in most distros as these are common technologies used in `/boot` nowadays, meaning the bootloader has to deal with them. So imagine A B filesystem images being replaced by btrfs subvolumes or by real lvm based partitions.

**Reply by Itxaka** [reply]
This would really mean a whole big change, although probably would be the better solution.

**Comment by Itxaka** [comment]
ok, goot enough feedback IMHO, this seems to be more work than expected with more changes than expected, so not worth it IMO.

**Comment by agracey** [comment]
What would be nice is to allow for components to be their own container layers. 

This would allow for smaller updates since if a middle layer gets updated, the node doesn't need to download the rest of the layers. 
It's my understanding that this should be possible and the container manifest can be rebuilt without rebuilding the layers or recomputing hashes (that said, the build service would need to support this)

**Comment by kkaempf** [comment]
That would require to build outside of openSUSE build service.
According to build service maintainers, container building (in build service) squashes all layers into one :-/

**Reply by agracey** [reply]
I'll try to argue that with the PMs as an enhancement in the future.
